[FEAT] Override Default Add friend, Teleport Messages

pierrecdevs opened the pull request:
This PR addresses enhancement / feature request of Issue https://github.com/ApertureViewer/Aperture-Viewer/issues/23 by allowing users to override default messages.


Addresses: https://github.com/ApertureViewer/Aperture-Viewer/issues/23

### Changelog

### Files changed

- `indra/newview/app_settings/settings_per_account.xml` - for the new override settings per user / account
- `indra/newview/skins/default/xui/en/notifications.xml` - notification dialogs 
- `indra/newview/skins/default/xui/en/strings.xml` - adding the default overrides.
- `indra/newview/skins/default/xui/en/panel_preferences_privacy.xml` - adds new panel/tab to privacy called **Overrides**
- `indra/newview/llavataractions.cpp` - overriding the actual messages in the string table / settings
- `indra/newview/llfloaterpreference.cpp` - added 3 new "changed" signals. `onOverrideAddFriendResponseChanged`, `onOverrideTeleportOfferResponseChanged` and `onOverrideTeleportRequestResponseChanged`
- `indra/newview/llfloaterpreference.h` - added the definitions for the  3 new "changed" signals. `onOverrideAddFriendResponseChanged`, `onOverrideTeleportOfferResponseChanged` and `onOverrideTeleportRequestResponseChanged`
- `indra/newview/llviewermessage.cpp` - overides the teleport offer message*

### New files 
- `indra/newview/skins/default/xui/en/panel_preferences_override.xml` - this is the new override panel


### *Known Issues / Caveats

Previously the `[REGION]` was translated to the current region. That does not happen if you change that as it's doing a string translation and I'm not sure if you can string translate of a string translation.... that would have to be tested...

williamweaver reviewed (changes requested):
Hi @pierrecdevs (and @afallenhope),

Thanks so much for this excellent PR! I've fetched the code, built it, and tested the functionality locally. I love it - the ability to customize these messages is a fantastic addition and works well! The UI integration is clean, and the core feature is solid. This directly addresses #23 and will be a welcome feature for many users.

During my code review, particularly looking at `indra/newview/app_settings/settings_per_account.xml`, I have one suggestion for a change before we move this to our formal QA process.

**Suggested Change: Prefix new settings keys with `AP`**

*(My apologies for not having this convention explicitly documented in our `CONTRIBUTING.md` sooner – that's an oversight on my part, and I'll be updating it to reflect this guidance for future contributions. I appreciate your understanding as we incorporate this for consistency.)*

To help us clearly identify Aperture-specific additions within the settings file, aid in future upstream merge management, and promote Aperture's unique features, could we please prefix the new settings keys with `AP` (for Aperture)?

This would apply to the following keys:
- `AddFriendReponseChanged` would become `APAddFriendReponseChanged`
- `AddFriendMessageResponse` would become `APAddFriendMessageResponse`
- `TeleportRequestResponseChanged` would become `APTeleportRequestResponseChanged`
- `TeleportRequestMessageResponse` would become `APTeleportRequestMessageResponse`
- `TeleportOfferResponseChanged` would become `APTeleportOfferResponseChanged`
- `TeleportOfferMessageResponse` would become `APTeleportOfferMessageResponse`

This would result in an XML block in `settings_per_account.xml` like this:

```xml
        <key>APAddFriendReponseChanged</key>
        <map>
        <key>Comment</key>
            <string>Does the Add Friend message differ from the default?</string>
        <key>Persist</key>
            <integer>1</integer>
        <key>Type</key>
            <string>Boolean</string>
        <key>Value</key>
            <integer>0</integer>
        </map>
        <key>APAddFriendMessageResponse</key>
        <map>
            <key>Comment</key>
            <string>Override the default add friend message</string>
            <key>Persist</key>
            <integer>1</integer>
        <key>Type</key>
            <string>String</string>
        <key>Value</key>
            <string/>
        </map>
        <key>APTeleportRequestResponseChanged</key>
        <map>
        <key>Comment</key>
            <string>Does the Teleport Request message differ from the default?</string>
        <key>Persist</key>
            <integer>1</integer>
        <key>Type</key>
            <string>Boolean</string>
        <key>Value</key>
            <integer>0</integer>
        </map>
        <key>APTeleportRequestMessageResponse</key>
        <map>
            <key>Comment</key>
            <string>Override the default teleport offer message</string>
            <key>Persist</key>
            <integer>1</integer>
        <key>Type</key>
            <string>String</string>
        <key>Value</key>
            <string/>
        </map>
        <key>APTeleportOfferResponseChanged</key>
        <map>
        <key>Comment</key>
            <string>Does the Teleport offer message differ from the default?</string>
        <key>Persist</key>
            <integer>1</integer>
        <key>Type</key>
            <string>Boolean</string>
        <key>Value</key>
            <integer>0</integer>
        </map>
       <key>APTeleportOfferMessageResponse</key>
        <map>
            <key>Comment</key>
            <string>Override the default teleport offer message</string>
            <key>Persist</key>
            <integer>1</integer>
        <key>Type</key>
            <string>String</string>
        <key>Value</key>
            <string/>
        </map>
 ```
 
This change to the setting key names would, of course, require corresponding updates in the C++ code where these settings are read and written.

Aside from this settings key naming suggestion, the feature implementation itself looks great from my initial developer review.

Please let me know your thoughts on this. If this change is agreeable and updated, I'll give it another quick check, and then we can proceed with drafting the detailed testing criteria for our QA team.

Again, fantastic work on this feature!

J1234-321 reviewed (commented):
Would like to see this dialog sized down and not expanded to whole screen.

https://gyazo.com/8e31a5d8b46c444111bd04f43f32cd22

Steps:
1.   In the Privacy/Overrides tab, add a long message to the max character limit and hit ok.
https://gyazo.com/8bf12c0576f69f22a35b56f14301a0f1
2.   Send an add friend request to a person.

Observed: The dialog stretches the entire width of the screen.  It would be nice if the dialog was a reasonable width and the text box be taller than a single line to show more of the message.  The height should allow the whole text box to show with no scroll bars.

https://gyazo.com/8e31a5d8b46c444111bd04f43f32cd22

pierrecdevs commented:
> Would like to see this dialog sized down and not expanded to whole screen.
> 
> https://gyazo.com/8e31a5d8b46c444111bd04f43f32cd22
> 
> Steps:
> 
> 1. In the Privacy/Overrides tab, add a long message to the max character limit and hit ok.
>    https://gyazo.com/8bf12c0576f69f22a35b56f14301a0f1
> 2. Send an add friend request to a person.
> 
> Observed: The dialog stretches the entire width of the screen. It would be nice if the dialog was a reasonable width and the text box be taller than a single line to show more of the message. The height should allow the whole text box to show with no scroll bars.
> 
> https://gyazo.com/8e31a5d8b46c444111bd04f43f32cd22

Hello @J1234-321 ,

Stating here for tracking purposes, as mentioned on Discord, the reason why the dialog stretches is due to how the `notifications.xml` dialog is displayed. It by defaults takes the length of the text and will attempt to "autosize". While investigating this, I was unable to find any way of overriding this without a "hacky" solution. While looking into the https://wiki.secondlife.com/wiki/Skinning_How_To/Window_sizing_and_resizing I have not found anything regarding how to properly override the notifications. 

TLDR; this is expected behaviour

pierrecdevs commented:
> Hi @pierrecdevs (and @afallenhope),
> 
> Thanks so much for this excellent PR! I've fetched the code, built it, and tested the functionality locally. I love it - the ability to customize these messages is a fantastic addition and works well! The UI integration is clean, and the core feature is solid. This directly addresses #23 and will be a welcome feature for many users.
> 
> During my code review, particularly looking at `indra/newview/app_settings/settings_per_account.xml`, I have one suggestion for a change before we move this to our formal QA process.
> 
> **Suggested Change: Prefix new settings keys with `AP`**
> 
> _(My apologies for not having this convention explicitly documented in our `CONTRIBUTING.md` sooner – that's an oversight on my part, and I'll be updating it to reflect this guidance for future contributions. I appreciate your understanding as we incorporate this for consistency.)_
> 
> To help us clearly identify Aperture-specific additions within the settings file, aid in future upstream merge management, and promote Aperture's unique features, could we please prefix the new settings keys with `AP` (for Aperture)?
> 
> This would apply to the following keys:
> 
> * `AddFriendReponseChanged` would become `APAddFriendReponseChanged`
> * `AddFriendMessageResponse` would become `APAddFriendMessageResponse`
> * `TeleportRequestResponseChanged` would become `APTeleportRequestResponseChanged`
> * `TeleportRequestMessageResponse` would become `APTeleportRequestMessageResponse`
> * `TeleportOfferResponseChanged` would become `APTeleportOfferResponseChanged`
> * `TeleportOfferMessageResponse` would become `APTeleportOfferMessageResponse`
> 
> This would result in an XML block in `settings_per_account.xml` like this:
> 
> ```
>         <key>APAddFriendReponseChanged</key>
>         <map>
>         <key>Comment</key>
>             <string>Does the Add Friend message differ from the default?</string>
>         <key>Persist</key>
>             <integer>1</integer>
>         <key>Type</key>
>             <string>Boolean</string>
>         <key>Value</key>
>             <integer>0</integer>
>         </map>
>         <key>APAddFriendMessageResponse</key>
>         <map>
>             <key>Comment</key>
>             <string>Override the default add friend message</string>
>             <key>Persist</key>
>             <integer>1</integer>
>         <key>Type</key>
>             <string>String</string>
>         <key>Value</key>
>             <string/>
>         </map>
>         <key>APTeleportRequestResponseChanged</key>
>         <map>
>         <key>Comment</key>
>             <string>Does the Teleport Request message differ from the default?</string>
>         <key>Persist</key>
>             <integer>1</integer>
>         <key>Type</key>
>             <string>Boolean</string>
>         <key>Value</key>
>             <integer>0</integer>
>         </map>
>         <key>APTeleportRequestMessageResponse</key>
>         <map>
>             <key>Comment</key>
>             <string>Override the default teleport offer message</string>
>             <key>Persist</key>
>             <integer>1</integer>
>         <key>Type</key>
>             <string>String</string>
>         <key>Value</key>
>             <string/>
>         </map>
>         <key>APTeleportOfferResponseChanged</key>
>         <map>
>         <key>Comment</key>
>             <string>Does the Teleport offer message differ from the default?</string>
>         <key>Persist</key>
>             <integer>1</integer>
>         <key>Type</key>
>             <string>Boolean</string>
>         <key>Value</key>
>             <integer>0</integer>
>         </map>
>        <key>APTeleportOfferMessageResponse</key>
>         <map>
>             <key>Comment</key>
>             <string>Override the default teleport offer message</string>
>             <key>Persist</key>
>             <integer>1</integer>
>         <key>Type</key>
>             <string>String</string>
>         <key>Value</key>
>             <string/>
>         </map>
> ```
> 
> This change to the setting key names would, of course, require corresponding updates in the C++ code where these settings are read and written.
> 
> Aside from this settings key naming suggestion, the feature implementation itself looks great from my initial developer review.
> 
> Please let me know your thoughts on this. If this change is agreeable and updated, I'll give it another quick check, and then we can proceed with drafting the detailed testing criteria for our QA team.
> 
> Again, fantastic work on this feature!

Updated!

J1234-321 reviewed (changes requested):
The overrides default to blank
https://gyazo.com/3c0b3fe72b5e718b17f44d85d0ac9bbc

Which allows the user to send blank friend requests and tp offers which is a change from the current behavior and may lead to user confusion.
https://gyazo.com/c989fd0978425973eeb5c2634b9ef5e0
https://gyazo.com/327e5d792a22f466202e6f8c6313ff11

Expected: 
The current defaults
https://gyazo.com/10321f5df903c0a9d6478813ff2eff60
https://gyazo.com/a1cd86f1daffe12015603d2e3f61541e

Note: Currently a request a teleport defaults to blank as well, so that default can stay blank.
https://gyazo.com/c6dbda83137ce52a356cfe3ac6a463af

afallenhope commented:
> The overrides default to blank https://gyazo.com/3c0b3fe72b5e718b17f44d85d0ac9bbc
> 
> Which allows the user to send blank friend requests and tp offers which is a change from the current behavior and may lead to user confusion. https://gyazo.com/c989fd0978425973eeb5c2634b9ef5e0 https://gyazo.com/327e5d792a22f466202e6f8c6313ff11
> 
> Expected: The current defaults https://gyazo.com/10321f5df903c0a9d6478813ff2eff60 https://gyazo.com/a1cd86f1daffe12015603d2e3f61541e
> 
> Note: Currently a request a teleport defaults to blank as well, so that default can stay blank. https://gyazo.com/c6dbda83137ce52a356cfe3ac6a463af

this has been resolved in https://github.com/ApertureViewer/Aperture-Viewer/pull/24/commits/2e7f91e21654195d9c12b1d5de69120ebc0df650 however, new bug somehow... I have to double check a few things. Thanks for pointing it out!

williamweaver reviewed (approved):
LGTM
